src/transformStateWithClones.js

```
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = state;
+  const history = [];
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        for (const key in action.extraData) {
+          newState[key] = action.extraData[key];
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      default:
+        break;
```

This violates checklist item #3: 'switch/case should always have default case for error handling'. The current `default` case does nothing, which means unknown actions are silently ignored. For proper error handling, you should consider throwing an error here to indicate that an invalid action type was provided.